### ISSUE TITLE
LIL sparse matrix described by faiss search index and distance?

### ISSUE THREAD

**Issue by lobpcg** [facebookresearch/faiss#1151]
# Summary

`distance_, index_ = index.search` 
implicitly determines a sparse matrix in LIL format with the indexes described by  `index_` and the corresponding matrix entries described by `distance_ `.

Is there a way to:

1. Explicitly define the LIL matrix without running the double `for` loops, like in the code below?
2. Or better yet,  explicitly define a function that multiplies an input vector or dense matrix `x` by this  matrix on the left, in some efficient way, directly using  `index_` and `distance_ ` without running the double `for` loops?

Is there already such a construct (1) or (2) in faiss built-in?

# Reproduction instructions

Python brute-force code to improve:
```
v_size = data.shape[1]
index = faiss.IndexFlatL2(v_size)
index.add(data)
m_size = index.ntotal

distance_, index_ = index.search(data, k)

a_ = lil_matrix((m_size, m_size))

for r in range(m_size):
    for c in arange(1, k):
        a_[r, index_[r, c]] = distance_[r, c]

def mult_by_a(x):
    return a_@x
```

**Comment by lobpcg** [comment]
I have found the answer myself:
```
a_ = lil_matrix((m_size, m_size))
a_.rows = index_[:, 1:]
a_.data = distance_[:, 1:]
```
seems to happily replace 
```
a_ = lil_matrix((m_size, m_size))
for r in range(m_size):
    for c in arange(1, k):
        a_[r, index_[r, c]] = distance_[r, c]
```

Closing

**Comment by nylocx** [comment]
Hi, I'm currently investigating the same use case, building a sparse matrix from the result of an NxN query with some k.
@lobpcg If I understand the docs right your code looks like it makes some assumptions that are most likely not always true. First is that you cut off the first element assuming that this always is the element itself, but if you have multiple identical vectors in your DB the order might swap, I have seen this with binary indices where `index_[i, 0]` was not `i` in a NxN query.
But his might be a corner case, what is more concerning is that the docs state that 
> `a_.rows` an array of rows, each of which is a sorted list of column indices of non-zero elements.
The sorted stuff is what is concerning me.

So I think you have to do an argsort on the `index_` along the last axis than use this to resort the `index_` and the `distances_` the same way. If my line of thinking is wrong please let me know where.

**Comment by lobpcg** [comment]
@nylocx sorry my example is incomplete - I should have specified what my "data" format was.  Could you please try:
```
from sklearn.datasets import make_moons
noise = 0.5
n_samples= 300
data, _ = make_moons(n_samples=n_samples, noise =noise)
data = data.astype('float32')
```
and confirm if my code runs with this data? 

I am quite sure that I have validated my code, giving the same matrix either way, so I believe that sorting is not needed, but please double-check. 

I do not work on this any longer, and cannot provide more help.

**Comment by nylocx** [comment]
@lobpcg I checked with some testing code and at first it seems that the sorting is really not needed, but you've piqued my curiosity. So I tried to dig deeper and the sorted property is used in the cython backend of the lil_matrix.
Here the bisect left and right functions are used
https://github.com/scipy/scipy/blob/master/scipy/sparse/_csparsetools.pyx.in#L313
see here:
https://github.com/scipy/scipy/blob/master/scipy/sparse/_csparsetools.pyx.in#L392

So there might be really strange bugs introduced by using the matrix as a sparse matrix and __changing__ content. 
So I think if you use your code and convert to a CSR matrix you can get off with it.

This is what I cocked up to show the "error" case:
``` python
from scipy.sparse import lil_matrix
import numpy as np
m = lil_matrix((5, 5))
rows = np.array([
    [2, 1],
    [3, 2],
    [4, 3],
    [0, 4],
    [1, 0],
]
)
data = np.array([
    [1, 2],
    [3, 4],
    [5, 6],
    [7, 8],
    [9, 10], 
])

m.rows = rows
m.data = data

m2 = m.tocsr().tolil()
m2[2, 3] = 11

m2.todense()

# This will fail with a ValueError
m[2, 3] = 13
```

If you comment the last line you see that it should work to assign an element in a sparse matrix but it fails if `rows` is not an array of lists but a 2d array. So I'm not 100% sure if I want to rely on the current behavior that read only access is working with a setup like this. So maybe reopen this issue and see if someone has a better idea to solve this conversion.

**Comment by nylocx** [comment]
After doing some more testing and benchmarking I came up with this:
```python
from scipy.sparse import lil_matrix
import numpy as np

index = np.tile(np.array([
    [1, 2],
    [0, 2],
    [0, 1],
]), (33333, 1))

distance = np.tile(np.array([
    [1, 2],
    [3, 4],
    [5, 6],
]), (33333, 1))

length, k = index.shape
```

```python
%%timeit
m1 = lil_matrix((length, length))
m1.rows = index
m1.data = distance
m1.tocsr()
```
> 324 ms ± 3.77 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

```python
%%timeit
length, k = index.shape
m2 = coo_matrix((distance.ravel(), (np.repeat(np.arange(length), k), index.ravel())), shape=(length, length))
m2.tocsr()
```
> 2.57 ms ± 13.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

```python
# Check that both solution result in equal matrices
(m1 != m2).nnz == 0
```
> True

So my conclusion is that constructing a `coo_matrix` and converting it to a `csr_matrix` with some relative simple reshaping  is much faster and does not misuse the data structure as the `lil_matrix` solution so for the future I will go with this. I hope that this will help others who have the same problem to make an elaborated decision which way to go.
